### ISSUE TITLE
Add error handling and logging to application

### ISSUE THREAD

**Issue by LukeParky** [GeospatialResearch/Digital-Twins#23]
When fatal errors occur the application should exit gracefully and provide a succinct message to users.
All errors should be logged to stderr.
Any time an exception occurs it must be handled in a way that does not just hide it, the code must either exit if it is fatal or it must pass the exception up the call stack if it is appropriate or the code must recover itself.

Check all try-except blocks for non-compliance.
I recommend that this issue be completed after or alongside #17 

Example of non-compliance from `src/dynamic_boundary_conditions/rain_depth_data_from_hirds.py:71`
```python
try:
        response = requests.get(url, headers=headers)
    except requests.exceptions.HTTPError as error:
        print("Request Failed", error)
    hirds_data = response.text
```

As you can see, if the `HTTPError` happens then it is caught and brushed under the covers and then immediately the program will crash after the try-except block finished

Another example: src/lidar/bg_flood_model.py:45
```python
    try:
        with open(rf"{file}\BG_param.txt", "w+") as file:
            file.write(f"topo = {dem_path}?{elev_var};\n")
            file.write(f"gpudevice = {gpudevice};\nmask = {mask};\ndx = {resolution};\n")
            file.write(f"smallnc = {smallnc};\n")
            file.write(f"outputtimestep = {outputtimestep};\nendtime = {endtime};\n")
            file.write(f"river = {river},{extents};\noutfile = {outfile};")
    except Exception as error:
        print(error, type(error))
        sys.exit()
````
here, the exception should be more specific than `Exception` and the error is printed, rather than handled. Then the system is closed by sys.exit(). It would be better to log to stderr and possibly allow the exception to raise so there is a chance to handle it.

These must be looked at on a case-by-case basis

**Comment by LukeParky** [comment]
As part of this the tests in tests/test_setup_environment/test_setup_environment.py should be reviewed and fixed if possib;e

**Comment by LukeParky** [comment]
Reopening because `test_setup_environment.py` has not been fixed

**Comment by LukeParky** [comment]
> Reopening because `test_setup_environment.py` has not been fixed

This doesn't actually have anything to do with this issue. Closed by #120
